src/merkle/smt/partial.rs

```diff
@@ -0,0 +1,312 @@
+    #[test]
+    fn partial_smt_insert_and_remove() {
+        let key0 = RpoDigest::from(Word::from(rand_array()));
+        let key1 = RpoDigest::from(Word::from(rand_array()));
+        let key2 = RpoDigest::from(Word::from(rand_array()));
+        // A key for which we won't add a value so it will be empty.
+        let key_empty = RpoDigest::from(Word::from(rand_array()));
+
+        let value0 = Word::from(rand_array());
+        let value1 = Word::from(rand_array());
+        let value2 = Word::from(rand_array());
+
+        let mut kv_pairs = vec![(key0, value0), (key1, value1), (key2, value2)];
+
+        // Add more random leaves.
+        kv_pairs.reserve(1000);
+        for _ in 0..1000 {
+            let key = RpoDigest::from(Word::from(rand_array()));
+            let value = Word::from(rand_array());
+            kv_pairs.push((key, value));
+        }
+
+        let mut full = Smt::with_entries(kv_pairs).unwrap();
+
+        // Constructing a partial SMT from proofs succeeds.
+        // ----------------------------------------------------------------------------------------
+
+        let proof0 = full.open(&key0);
+        let proof2 = full.open(&key2);
+        let proof_empty = full.open(&key_empty);
+
+        assert!(proof_empty.leaf().is_empty());
+
+        let mut partial = PartialSmt::from_merkle_paths([
+            proof0.into_parts(),
+            proof2.into_parts(),
+            proof_empty.into_parts(),
+        ])
+        .unwrap();
+
+        assert_eq!(full.root(), partial.root());
+        assert_eq!(partial.get_value(&key0), value0);
+        assert_eq!(partial.get_value(&key1), EMPTY_WORD);
+        assert_eq!(partial.get_value(&key2), value2);
+
+        // Insert new values for added keys with empty and non-empty values.
+        // ----------------------------------------------------------------------------------------
+
+        let new_value0 = Word::from(rand_array());
+        let new_value2 = Word::from(rand_array());
+        // A non-empty value for the key that was previously empty.
+        let new_value_empty_key = Word::from(rand_array());
+
+        full.insert(key0, new_value0);
+        full.insert(key2, new_value2);
+        full.insert(key_empty, new_value_empty_key);
+
+        partial.insert(key0, new_value0).unwrap();
+        partial.insert(key2, new_value2).unwrap();
+        // This updates a key whose value was previously empty.
+        partial.insert(key_empty, new_value_empty_key).unwrap();
+
+        assert_eq!(full.root(), partial.root());
+        assert_eq!(partial.get_value(&key0), new_value0);
+        assert_eq!(partial.get_value(&key2), new_value2);
+        assert_eq!(partial.get_value(&key_empty), new_value_empty_key);
+
+        // Remove an added key.
+        // ----------------------------------------------------------------------------------------
+
+        full.insert(key0, EMPTY_WORD);
+        partial.insert(key0, EMPTY_WORD).unwrap();
```

Comment:
Inserting `EMPTY_WORD` effectively removes the value from the partial SMT, right? That is, once we insert `EMPTY_WORD` for a key, we won't be able to update this key's value again (i.e., after this line, `partial.insert(key0, some_new_value).unwrap();` would fail) - right?

This is not an issue for this PR - but maybe we should make this behavior more explicit and also we could create an issue to see if we can improve this in the future.

Reply:
Yes, that is the behavior. This was relatively easy to fix (reinserting an `SmtLeaf::Empty` when inserting `value == EMPTY_WORD` to consider the leaf still tracked), so I made the change.